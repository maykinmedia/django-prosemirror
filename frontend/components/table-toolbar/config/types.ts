import { IconKeys } from "@/plugins/icons";
import { MenuItemSpec } from "prosemirror-menu";
import { Command } from "prosemirror-state";

export interface TableToolbarButton {
    icon?: IconKeys;
    title: string;
    command: Command;
    run?: Command;
    isActive?: MenuItemSpec["active"];
}

export interface TableToolbarDropdown {
    icon: IconKeys;
    title: string;
    items: TableToolbarButton[];
}

export interface TableToolbarPosition {
    top: number;
    left: number;
}

export interface TableElementData {
    tableElement: Element;
    tableWrapper: Element;
}
