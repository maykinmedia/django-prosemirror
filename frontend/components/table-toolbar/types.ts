import { IconKeys } from "@/plugins/icons";
import { EditorState, Transaction } from "prosemirror-state";
import { EditorView } from "prosemirror-view";

export interface IToolbarMenuItem {
    icon?: IconKeys;
    title: string;
    command: (
        state: EditorState,
        dispatch?: (tr: Transaction) => void,
    ) => boolean;
    run?: (state: EditorState, dispatch?: (tr: Transaction) => void) => boolean;
    isActive?: (view: EditorView) => boolean;
}

export type ButtonOrDropdown = ButtonOptions & {
    items?: IToolbarMenuItem[];
};

export interface ButtonOptions {
    title?: string;
    class?: string;
    icon?: IconKeys;
    disabled?: boolean;
    command?: (
        state: EditorState,
        dispatch?: (tr: Transaction) => void,
    ) => boolean;
    isActive?: (view: EditorView) => boolean;
}
