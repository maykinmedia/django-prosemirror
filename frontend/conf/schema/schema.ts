import {
    ClassMapping,
    MarkDefinition,
    NodeDefinition,
} from "@/conf/schema/abstract";
import {
    CodeMark,
    ItalicMark,
    LinkMark,
    StrikeThroughMark,
    StrongMark,
    UnderlineMark,
} from "@/conf/schema/marks";
import {
    BlockQuoteNode,
    BulletListNode,
    CodeBlockNode,
    HardBreakNode,
    HeadingNode,
    HorizontalRuleNode,
    ImageNode,
    ListItemNode,
    OrderedListNode,
    ParagraphNode,
    TableCellNode,
    TableHeaderNode,
    TableNode,
    TableRowNode,
} from "@/conf/schema/nodes";
import { IDPMSettings, MarkType, NodeType } from "@/conf/types";
import { MarkSpec, NodeSpec, Schema } from "prosemirror-model";

export class DPMSchema {
    settings: IDPMSettings;
    classMapping?: ClassMapping;

    constructor(settings: IDPMSettings) {
        this.settings = settings;
        try {
            this._validateSchema();
        } catch (e) {
            console.error(e);
        }
    }

    // Get the specs of all the nodes.
    _createNodes(): Record<NodeType, NodeDefinition> {
        return {
            // Required nodes
            [NodeType.DOC]: { spec: { content: "block+" } } as NodeDefinition,
            [NodeType.TEXT]: { spec: { group: "inline" } } as NodeDefinition,
            [NodeType.PARAGRAPH]: new ParagraphNode(this.classMapping!),

            // Default
            [NodeType.HEADING]: new HeadingNode(this.classMapping!),
            [NodeType.BLOCKQUOTE]: new BlockQuoteNode(this.classMapping!),
            [NodeType.IMAGE]: new ImageNode(this.classMapping!),
            [NodeType.HORIZONTAL_RULE]: new HorizontalRuleNode(
                this.classMapping!,
            ),
            [NodeType.CODE_BLOCK]: new CodeBlockNode(this.classMapping!),
            [NodeType.HARD_BREAK]: new HardBreakNode(this.classMapping!),

            // List
            [NodeType.BULLET_LIST]: new BulletListNode(this.classMapping!),
            [NodeType.ORDERED_LIST]: new OrderedListNode(this.classMapping!),
            [NodeType.LIST_ITEM]: new ListItemNode(this.classMapping!),

            // Table
            [NodeType.TABLE]: new TableNode(this.classMapping!),
            [NodeType.TABLE_CELL]: new TableCellNode(this.classMapping!),
            [NodeType.TABLE_HEADER]: new TableHeaderNode(this.classMapping!),
            [NodeType.TABLE_ROW]: new TableRowNode(this.classMapping!),
        };
    }

    _getNodes(): Record<NodeType, NodeSpec> {
        const nodes = this._createNodes();

        const requiredNodes = {
            [NodeType.DOC]: nodes[NodeType.DOC].spec,
            [NodeType.PARAGRAPH]: nodes[NodeType.PARAGRAPH].spec,
            [NodeType.TEXT]: nodes[NodeType.TEXT].spec,
        } as Record<NodeType, NodeSpec>;

        return this.settings.allowedNodes?.reduce((acc, cur) => {
            // Append node
            if (nodes[cur]) acc[cur] = nodes[cur].spec;

            // Execute some commands to make sure the table is set correct.
            if (cur === NodeType.TABLE) {
                document.execCommand?.("enableObjectResizing", false, "false");
                document.execCommand?.(
                    "enableInlineTableEditing",
                    false,
                    "false",
                );
            }

            // Append list_item if bullet/ordered list is appended
            if (NodeType.ORDERED_LIST === cur || NodeType.BULLET_LIST === cur)
                acc[NodeType.LIST_ITEM] = nodes[NodeType.LIST_ITEM].spec;
            return acc;
        }, requiredNodes);
    }

    // Helper method to get all mark specs
    _createMarks(): Record<MarkType, MarkDefinition> {
        return {
            [MarkType.LINK]: new LinkMark(this.classMapping!),
            [MarkType.STRONG]: new StrongMark(this.classMapping!),
            [MarkType.ITALIC]: new ItalicMark(this.classMapping!),
            [MarkType.CODE]: new CodeMark(this.classMapping!),
            [MarkType.STRIKETHROUGH]: new StrikeThroughMark(this.classMapping!),
            [MarkType.UNDERLINE]: new UnderlineMark(this.classMapping!),
        };
    }

    _getMarks(): Record<MarkType, MarkSpec> {
        const marks = this._createMarks();

        return this.settings.allowedMarks?.reduce(
            (acc, cur) => {
                if (marks[cur]) acc[cur] = marks[cur].spec;
                return acc;
            },
            {} as Record<MarkType, MarkSpec>,
        );
    }
    /**
     * Validate the schema to check if all nodes/marks are valid.
     * @throws {Error} no paragraph included
     * @throws {TypeError} Unsupported nodes/marks inside the schema or allowedNodes/allowedMarks is not an iterable
     */
    _validateSchema() {
        // Validate node types
        if (!this.settings.allowedNodes?.includes(NodeType.PARAGRAPH)) {
            throw new Error(
                `Your list of allowed node types must inlcude ${NodeType.PARAGRAPH}`,
            );
        }

        try {
            for (const nodeType of this.settings.allowedNodes) {
                if (!Object.values(NodeType).includes(nodeType)) {
                    throw new TypeError(
                        `Expected NodeType enum, got ${typeof nodeType}: ${nodeType}`,
                    );
                }
            }
        } catch (error) {
            if (error instanceof TypeError) {
                throw error;
            }
            throw new TypeError(
                "allowedNodeTypes must be an iterable of NodeType enums",
            );
        }

        // Validate mark types
        try {
            for (const markType of this.settings.allowedMarks) {
                if (!Object.values(MarkType).includes(markType)) {
                    throw new TypeError(
                        `Expected MarkType enum, got ${typeof markType}: ${markType}`,
                    );
                }
            }
        } catch (error) {
            if (error instanceof TypeError) {
                throw error;
            }
            throw new TypeError(
                "allowedMarkTypes must be an iterable of MarkType enums",
            );
        }
    }

    /**
     * Validate
     */
    get schema() {
        this.classMapping = new ClassMapping(this.settings?.classNames ?? {});

        return new Schema({
            nodes: this._getNodes(),
            marks: this._getMarks(),
        });
    }
}
