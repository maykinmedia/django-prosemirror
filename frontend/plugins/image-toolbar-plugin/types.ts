// This file is kept for backward compatibility but ImageToolbarConfig is no longer used
// The image toolbar now only accepts ButtonOrDropdown[] directly
