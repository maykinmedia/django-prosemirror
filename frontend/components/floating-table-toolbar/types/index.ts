import { EditorState, Transaction } from "prosemirror-state";

export interface FloatingTableToolbarButton {
    icon?: string;
    title: string;
    command: (
        state: EditorState,
        dispatch?: (tr: Transaction) => void,
    ) => boolean;
    run?: (state: EditorState, dispatch?: (tr: Transaction) => void) => boolean;
    isActive?: (state: EditorState) => boolean;
}

export interface FloatingTableToolbarDropdown {
    icon: string;
    title: string;
    items: FloatingTableToolbarButton[];
}

export interface ToolbarPosition {
    top: number;
    left: number;
}

export interface TableElementData {
    tableElement: Element;
    tableWrapper: Element;
}
