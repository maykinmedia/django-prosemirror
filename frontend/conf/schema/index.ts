export * from "@/conf/schema/abstract";
export * from "@/conf/schema/marks";
export * from "@/conf/schema/nodes";
export * from "@/conf/schema/schema";
