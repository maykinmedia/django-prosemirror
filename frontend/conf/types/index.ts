/**
 * Type definitions for the schema module.
 *
 * These enums contains the key-value pairs of the available nodes/marks - Each key MUST
 * contain the same value as defined in `django_prosemirror/schema/types.py`
 */

export enum NodeType {
    /** Enumeration of node types that can be used in Prosemirror schemas. */
    // Required nodes (always present)
    DOC = "doc",
    TEXT = "text",
    PARAGRAPH = "paragraph",
    // Optional nodes
    BLOCKQUOTE = "blockquote",
    HORIZONTAL_RULE = "horizontal_rule",
    HEADING = "heading",
    IMAGE = "image",
    HARD_BREAK = "hard_break",
    CODE_BLOCK = "code_block",
    // List nodes
    BULLET_LIST = "bullet_list",
    ORDERED_LIST = "ordered_list",
    LIST_ITEM = "list_item",
    // Table nodes
    TABLE = "table",
    TABLE_ROW = "table_row",
    TABLE_CELL = "table_cell",
    TABLE_HEADER = "table_header",
}

export enum MarkType {
    /** Enumeration of mark types that can be used in Prosemirror schemas. */
    STRONG = "strong",
    ITALIC = "em",
    LINK = "link",
    CODE = "code",
    UNDERLINE = "underline",
    STRIKETHROUGH = "strikethrough",
}

export interface IDPMSettings {
    /** Set to false to disable the menu bar. */
    menubar?: boolean;
    /** Set to false to disable the history plugin.*/
    history?: boolean;
    /** Set to false to make the menu bar non-floating. */
    floatingMenu?: boolean;
    language?: LanguageCodeEnum | string;
    debug?: boolean;
    classNames?: Record<string, string>;
    allowedNodes: Array<NodeType>;
    allowedMarks: Array<MarkType>;
}

export enum LanguageCodeEnum {
    NL = "nl",
    EN = "en",
}

export interface IDPMTranslations {
    "Join with above block": string;
    "Lift out of enclosing block": string;
    "Select parent node": string;
    "Undo last change": string;
    "Redo last undone change": string;
    "Insert image": string;
    Image: string;
    Location: string;
    Title: string;
    Description: string;
    "Add or remove link": string;
    "Create a link": string;
    "Link target": string;
    "Toggle strong style": string;
    "Toggle emphasis": string;
    "Toggle code font": string;
    "Wrap in bullet list": string;
    "Wrap in ordered list": string;
    "Wrap in block quote": string;
    "Change to paragraph": string;
    Plain: string;
    "Change to code block": string;
    Code: string;
    "Change to heading 1": string;
    "Change to heading 2": string;
    "Change to heading 3": string;
    "Change to heading 4": string;
    "Change to heading 5": string;
    "Change to heading 6": string;
    "Level 1": string;
    "Level 2": string;
    "Level 3": string;
    "Level 4": string;
    "Level 5": string;
    "Level 6": string;
    "Insert horizontal rule": string;
    "Type...": string;
    "Toggle underline": string;
    "Toggle strikethrough": string;
    "Horizontal rule": string;
    OK: string;
    Cancel: string;
    "Delete table": string;
    "Add column after": string;
    "Add column before": string;
    "Delete column": string;
    "Toggle header column": string;
    "Merge cells": string;
    "Split cell": string;
    "Add row before": string;
    "Add row after": string;
    "Delete row": string;
    "Toggle header row": string;
    "Insert table": string;
    "Row operations": string;
    "Column operations": string;
    "Cell operations": string;
}
