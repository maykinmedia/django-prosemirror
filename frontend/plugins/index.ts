import { IDPMSettings } from "@/conf/types";
import { baseKeymap } from "prosemirror-commands";
import { dropCursor } from "prosemirror-dropcursor";
import { buildInputRules, buildKeymap } from "prosemirror-example-setup";
import { gapCursor } from "prosemirror-gapcursor";
import { history } from "prosemirror-history";
import { keymap } from "prosemirror-keymap";
import { menuBar } from "prosemirror-menu";
import { Schema } from "prosemirror-model";
import { Plugin } from "prosemirror-state";
import { buildMenuItems } from "./menubar";
import { tablePlugins } from "./table";

export interface DjangoProsemirrorSetup {
    /** The model schema of the editor. */
    schema: Schema;
    /** Set to false to disable the menu bar. */
    menubar?: boolean;
    /** Set to false to disable the history plugin.*/
    history?: boolean;
    /** Set to false to make the menu bar non-floating. */
    floatingMenu?: boolean;
}

/**
 * Get the plugins for the django prosemirror editor.
 * @param settings
 * @returns The array of
 */
export function getDPMPlugins(
    schema: Schema,
    settings?: IDPMSettings,
): (Plugin & {
    type?: string;
    config?: { content: unknown };
    schema?: unknown;
})[] {
    const plugins = [
        /**
         * Table plugin for table cell editing and a floating menubar.
         */
        ...tablePlugins(schema),
        /**
         * Input rules for smart quotes and creating the block types in the schema
         * using markdown conventions.
         */
        buildInputRules(schema),
        /**
         * A keymap that defines keys to create and manipulate the nodes in the schema
         */
        keymap(buildKeymap(schema)),
        /**
         * A keymap binding the default keys provided by the
         * prosemirror-commands module
         */
        keymap(baseKeymap),
        /**
         * The drop cursor plugin
         */
        dropCursor(),
        /**
         * The gap cursor plugin
         */
        gapCursor(),
        /**
         * The menubar plugin
         */
        menuBar({
            // floating: false,
            // floating: settings?.floatingMenu !== false,
            content: buildMenuItems(schema, settings?.history).fullMenu,
        }),
        /**
         * The undo history plugin
         */
        history(),
    ];

    // TODO once we add (more) settings include this part so the menubar/history can be turned off.
    // if (settings.menubar !== false)
    //     plugins.push(
    //         menuBar({
    //             floating: settings.floatingMenu !== false,
    //             content: buildMenuItems(settings.schema, settings.history)
    //                 .fullMenu,
    //         })
    //     );

    // if (settings.history !== false) plugins.push(history());

    return plugins;
}
