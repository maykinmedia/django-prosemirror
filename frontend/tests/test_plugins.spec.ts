import { describe, it, expect, vi, beforeEach } from "vitest";
import { getDjangoProsemirrorPlugins } from "../plugins/index";
import { DjangoProsemirrorSettings, LanguageCodeEnum } from "../types/types";
import { Schema } from "prosemirror-model";
import { SchemaNodesEnum } from "@/schema/choices";

// Mock the ProseMirror plugins
vi.mock("prosemirror-keymap", () => ({
    keymap: vi.fn((keymap) => ({ type: "keymap", keymap })),
}));

vi.mock("prosemirror-history", () => ({
    history: vi.fn(() => ({ type: "history" })),
}));

vi.mock("prosemirror-commands", () => ({
    baseKeymap: { "Mod-z": "undo", "Mod-y": "redo" },
}));

vi.mock("prosemirror-dropcursor", () => ({
    dropCursor: vi.fn(() => ({ type: "dropCursor" })),
}));

vi.mock("prosemirror-gapcursor", () => ({
    gapCursor: vi.fn(() => ({ type: "gapCursor" })),
}));

vi.mock("prosemirror-menu", () => ({
    menuBar: vi.fn((config) => ({ type: "menuBar", config })),
}));

vi.mock("prosemirror-example-setup", () => ({
    buildInputRules: vi.fn((schema) => ({ type: "inputRules", schema })),
    buildKeymap: vi.fn((schema) => ({ "Mod-b": "toggleStrong", schema })),
}));

vi.mock("../plugins/menubar/index.ts", () => ({
    buildMenuItems: vi.fn((schema, history) => ({
        fullMenu: [
            [{ title: "Bold" }, { title: "Italic" }],
            [{ title: "Heading" }],
        ],
        schema,
        history,
    })),
}));

describe("plugins/index", () => {
    let mockSchema: Schema;

    beforeEach(() => {
        vi.clearAllMocks();

        // Create a simple mock schema
        mockSchema = {
            nodes: {
                doc: { name: "doc" },
                paragraph: { name: "paragraph" },
                text: { name: "text" },
            },
            marks: {
                strong: { name: "strong" },
                em: { name: "em" },
            },
        } as unknown as Schema;
    });

    describe("getDjangoProsemirrorPlugins", () => {
        it("should return array of plugins with default settings", () => {
            const plugins = getDjangoProsemirrorPlugins(mockSchema);

            expect(Array.isArray(plugins)).toBe(true);
            expect(plugins).toMatchSnapshot(); // 7 plugins by default
        });

        it("should include input rules plugin", () => {
            const plugins = getDjangoProsemirrorPlugins(mockSchema);
            const inputRulesPlugin = plugins.find(
                (p) => p.type === "inputRules",
            );

            expect(inputRulesPlugin).toMatchSnapshot();
            expect(inputRulesPlugin?.schema).toBe(mockSchema);
        });

        it("should include keymap plugins", () => {
            const plugins = getDjangoProsemirrorPlugins(mockSchema);
            const keymapPlugins = plugins.filter((p) => p.type === "keymap");

            expect(keymapPlugins).toHaveLength(2); // Custom keymap + base keymap
        });

        it("should include drop cursor plugin", () => {
            const plugins = getDjangoProsemirrorPlugins(mockSchema);
            const dropCursorPlugin = plugins.find(
                (p) => p.type === "dropCursor",
            );

            expect(dropCursorPlugin).toMatchSnapshot();
        });

        it("should include gap cursor plugin", () => {
            const plugins = getDjangoProsemirrorPlugins(mockSchema);
            const gapCursorPlugin = plugins.find((p) => p.type === "gapCursor");

            expect(gapCursorPlugin).toMatchSnapshot();
        });

        it("should include menu bar plugin", () => {
            const plugins = getDjangoProsemirrorPlugins(mockSchema);
            const menuBarPlugin = plugins.find((p) => p.type === "menuBar");

            expect(menuBarPlugin).toMatchSnapshot();
        });

        it("should include history plugin", () => {
            const plugins = getDjangoProsemirrorPlugins(mockSchema);
            const historyPlugin = plugins.find((p) => p.type === "history");

            expect(historyPlugin).toMatchSnapshot();
        });

        it("should pass settings to buildMenuItems", () => {
            const settings: DjangoProsemirrorSettings = {
                allowedNodes: [
                    SchemaNodesEnum.HEADING,
                    SchemaNodesEnum.PARAGRAPH,
                ],
                history: false,
                language: LanguageCodeEnum.EN,
            };

            const result = getDjangoProsemirrorPlugins(mockSchema, settings);

            // Just verify that plugins are created without errors
            expect(result).toMatchSnapshot();
            expect(Array.isArray(result)).toBe(true);
        });

        it("should pass default history setting when not specified", () => {
            const result = getDjangoProsemirrorPlugins(mockSchema);

            // Just verify that plugins are created without errors
            expect(result).toMatchSnapshot();
            expect(Array.isArray(result)).toBe(true);
        });

        it("should work with minimal schema", () => {
            const minimalSchema = {
                nodes: { doc: {}, paragraph: {}, text: {} },
                marks: {},
            } as unknown as Schema;

            const plugins = getDjangoProsemirrorPlugins(minimalSchema);

            expect(Array.isArray(plugins)).toBe(true);
            expect(plugins.length).toBeGreaterThan(0);
        });

        it("should handle undefined settings gracefully", () => {
            expect(() => {
                getDjangoProsemirrorPlugins(mockSchema, undefined);
            }).not.toThrow();
        });

        it("should handle empty settings object", () => {
            const settings: DjangoProsemirrorSettings = {
                allowedNodes: [],
            };

            expect(() => {
                getDjangoProsemirrorPlugins(mockSchema, settings);
            }).not.toThrow();
        });

        it("should maintain plugin order", () => {
            const plugins = getDjangoProsemirrorPlugins(mockSchema);
            const pluginTypes = plugins.map((p) => p.type);

            expect(pluginTypes).toEqual([
                "inputRules",
                "keymap", // buildKeymap
                "keymap", // baseKeymap
                "dropCursor",
                "gapCursor",
                "menuBar",
                "history",
            ]);
        });

        it("should pass correct parameters to each plugin factory", () => {
            const plugins = getDjangoProsemirrorPlugins(mockSchema);

            // Verify that all expected plugin types are present
            const pluginTypes = plugins.map((p) => p.type);
            expect(pluginTypes).toContain("inputRules");
            expect(pluginTypes).toContain("keymap");
            expect(pluginTypes).toContain("dropCursor");
            expect(pluginTypes).toContain("gapCursor");
            expect(pluginTypes).toContain("menuBar");
            expect(pluginTypes).toContain("history");
        });

        it("should create menu bar with correct content structure", () => {
            const plugins = getDjangoProsemirrorPlugins(mockSchema);
            const menuBarPlugin = plugins.find((p) => p.type === "menuBar");

            expect(menuBarPlugin).toMatchSnapshot();
            expect(menuBarPlugin?.config?.content).toMatchSnapshot();
        });
    });

    describe("Plugin configuration", () => {
        it("should configure plugins with different settings", () => {
            const settings1: DjangoProsemirrorSettings = {
                allowedNodes: [SchemaNodesEnum.HEADING],
                history: true,
            };
            const settings2: DjangoProsemirrorSettings = {
                allowedNodes: [SchemaNodesEnum.PARAGRAPH],
                history: false,
            };

            const plugins1 = getDjangoProsemirrorPlugins(mockSchema, settings1);
            const plugins2 = getDjangoProsemirrorPlugins(mockSchema, settings2);

            // Both should return the same structure
            expect(plugins1).toEqual(plugins2);
        });

        it("should handle complex settings object", () => {
            const complexSettings: DjangoProsemirrorSettings = {
                allowedNodes: [
                    SchemaNodesEnum.HEADING,
                    SchemaNodesEnum.PARAGRAPH,
                    SchemaNodesEnum.BLOCKQUOTE,
                    SchemaNodesEnum.UNORDERED_LIST,
                ],
                menubar: true,
                history: true,
                floatingMenu: true,
                language: LanguageCodeEnum.NL,
                debug: true,
                classNames: {
                    strong: "custom-strong",
                    em: "custom-em",
                },
            };

            expect(() => {
                getDjangoProsemirrorPlugins(mockSchema, complexSettings);
            }).not.toThrow();
        });
    });
});
