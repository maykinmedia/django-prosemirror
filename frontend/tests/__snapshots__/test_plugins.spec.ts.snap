// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugins/index > getDjangoProsemirrorPlugins > should create menu bar with correct content structure 1`] = `
{
  "config": {
    "content": [
      [
        {
          "title": "Bold",
        },
        {
          "title": "Italic",
        },
      ],
      [
        {
          "title": "Heading",
        },
      ],
    ],
  },
  "type": "menuBar",
}
`;

exports[`plugins/index > getDjangoProsemirrorPlugins > should create menu bar with correct content structure 2`] = `
[
  [
    {
      "title": "Bold",
    },
    {
      "title": "Italic",
    },
  ],
  [
    {
      "title": "Heading",
    },
  ],
]
`;

exports[`plugins/index > getDjangoProsemirrorPlugins > should include drop cursor plugin 1`] = `
{
  "type": "dropCursor",
}
`;

exports[`plugins/index > getDjangoProsemirrorPlugins > should include gap cursor plugin 1`] = `
{
  "type": "gapCursor",
}
`;

exports[`plugins/index > getDjangoProsemirrorPlugins > should include history plugin 1`] = `
{
  "type": "history",
}
`;

exports[`plugins/index > getDjangoProsemirrorPlugins > should include input rules plugin 1`] = `
{
  "schema": {
    "marks": {
      "em": {
        "name": "em",
      },
      "strong": {
        "name": "strong",
      },
    },
    "nodes": {
      "doc": {
        "name": "doc",
      },
      "paragraph": {
        "name": "paragraph",
      },
      "text": {
        "name": "text",
      },
    },
  },
  "type": "inputRules",
}
`;

exports[`plugins/index > getDjangoProsemirrorPlugins > should include menu bar plugin 1`] = `
{
  "config": {
    "content": [
      [
        {
          "title": "Bold",
        },
        {
          "title": "Italic",
        },
      ],
      [
        {
          "title": "Heading",
        },
      ],
    ],
  },
  "type": "menuBar",
}
`;

exports[`plugins/index > getDjangoProsemirrorPlugins > should pass default history setting when not specified 1`] = `
[
  {
    "schema": {
      "marks": {
        "em": {
          "name": "em",
        },
        "strong": {
          "name": "strong",
        },
      },
      "nodes": {
        "doc": {
          "name": "doc",
        },
        "paragraph": {
          "name": "paragraph",
        },
        "text": {
          "name": "text",
        },
      },
    },
    "type": "inputRules",
  },
  {
    "keymap": {
      "Mod-b": "toggleStrong",
      "schema": {
        "marks": {
          "em": {
            "name": "em",
          },
          "strong": {
            "name": "strong",
          },
        },
        "nodes": {
          "doc": {
            "name": "doc",
          },
          "paragraph": {
            "name": "paragraph",
          },
          "text": {
            "name": "text",
          },
        },
      },
    },
    "type": "keymap",
  },
  {
    "keymap": {
      "Mod-y": "redo",
      "Mod-z": "undo",
    },
    "type": "keymap",
  },
  {
    "type": "dropCursor",
  },
  {
    "type": "gapCursor",
  },
  {
    "config": {
      "content": [
        [
          {
            "title": "Bold",
          },
          {
            "title": "Italic",
          },
        ],
        [
          {
            "title": "Heading",
          },
        ],
      ],
    },
    "type": "menuBar",
  },
  {
    "type": "history",
  },
]
`;

exports[`plugins/index > getDjangoProsemirrorPlugins > should pass settings to buildMenuItems 1`] = `
[
  {
    "schema": {
      "marks": {
        "em": {
          "name": "em",
        },
        "strong": {
          "name": "strong",
        },
      },
      "nodes": {
        "doc": {
          "name": "doc",
        },
        "paragraph": {
          "name": "paragraph",
        },
        "text": {
          "name": "text",
        },
      },
    },
    "type": "inputRules",
  },
  {
    "keymap": {
      "Mod-b": "toggleStrong",
      "schema": {
        "marks": {
          "em": {
            "name": "em",
          },
          "strong": {
            "name": "strong",
          },
        },
        "nodes": {
          "doc": {
            "name": "doc",
          },
          "paragraph": {
            "name": "paragraph",
          },
          "text": {
            "name": "text",
          },
        },
      },
    },
    "type": "keymap",
  },
  {
    "keymap": {
      "Mod-y": "redo",
      "Mod-z": "undo",
    },
    "type": "keymap",
  },
  {
    "type": "dropCursor",
  },
  {
    "type": "gapCursor",
  },
  {
    "config": {
      "content": [
        [
          {
            "title": "Bold",
          },
          {
            "title": "Italic",
          },
        ],
        [
          {
            "title": "Heading",
          },
        ],
      ],
    },
    "type": "menuBar",
  },
  {
    "type": "history",
  },
]
`;

exports[`plugins/index > getDjangoProsemirrorPlugins > should return array of plugins with default settings 1`] = `
[
  {
    "schema": {
      "marks": {
        "em": {
          "name": "em",
        },
        "strong": {
          "name": "strong",
        },
      },
      "nodes": {
        "doc": {
          "name": "doc",
        },
        "paragraph": {
          "name": "paragraph",
        },
        "text": {
          "name": "text",
        },
      },
    },
    "type": "inputRules",
  },
  {
    "keymap": {
      "Mod-b": "toggleStrong",
      "schema": {
        "marks": {
          "em": {
            "name": "em",
          },
          "strong": {
            "name": "strong",
          },
        },
        "nodes": {
          "doc": {
            "name": "doc",
          },
          "paragraph": {
            "name": "paragraph",
          },
          "text": {
            "name": "text",
          },
        },
      },
    },
    "type": "keymap",
  },
  {
    "keymap": {
      "Mod-y": "redo",
      "Mod-z": "undo",
    },
    "type": "keymap",
  },
  {
    "type": "dropCursor",
  },
  {
    "type": "gapCursor",
  },
  {
    "config": {
      "content": [
        [
          {
            "title": "Bold",
          },
          {
            "title": "Italic",
          },
        ],
        [
          {
            "title": "Heading",
          },
        ],
      ],
    },
    "type": "menuBar",
  },
  {
    "type": "history",
  },
]
`;
